### PR TITLE
adds html tooltip, adds a tooltip to character pref about naming policy

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
* adds html tooltip
* adds a tooltip to character pref about naming policy (as an example for how to use this)

## Why It's Good For The Game
tooltip is good to implement stuff
also, notifying players about naming policy is good (but it should have a link to our wiki, I think)

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

</details>

## Changelog
:cl:
code: html tooltip helping DEFINEs and proc
add: character pref will now notify about naming policy
config: add configurable tooltip config file
/:cl:

### CODE REVIEW

File: code/modules/client/preferences.dm [hunk 1]
```diff
@@ -169,7 +169,7 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			dat += "<a href='?_src_=prefs;preference=name;task=random'>Random Name</A> "
 			dat += "<a href='?_src_=prefs;preference=name'>Always Random Name: [active_character.be_random_name ? "Yes" : "No"]</a><BR>"
 
-			dat += "<b>Name:</b> "
+			dat += "<b>[TOOLTIP_WRAPPER("Name:", 400, "Each speices must follow a naming policy. Please check our \"Naming Guidelines\" in our wiki.")]</b> "
```
Comment (itsmeow) [inline code/modules/client/preferences.dm]: Please read the "Naming Guidelines" page on the wiki and Rule 1.7.

File: code/__DEFINES/html_assistant.dm [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  \
+  position: absolute;\
+  z-index: 1;\
+}\
+\
+.tooltip:hover .tooltiptext {\
+  visibility: visible;\
+}\
+</style>"
+// I tried `width: max-content;` in css, but it's not working. giving custom width_px only works for now.
```
Comment (itsmeow) [inline code/__DEFINES/html_assistant.dm]: IE11 does not support the max-content attribute.

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/__DEFINES/html_assistant.dm (hunk 2 above)
Comment (EvilDragonfiend) [reply code/__DEFINES/html_assistant.dm]: that explains why

**Comment by itsmeow** [comment]
If you can make it actually link to both pages it would be very good

**Comment by EvilDragonfiend** [comment]
> If you can make it actually link to both pages it would be very good

I tried, but I have no idea how I can make a href link to pop up a wiki page.

**Comment by itsmeow** [comment]
basic html `<a href="link">Text</a>`

File: code/modules/client/preferences.dm [hunk 3]
```diff
@@ -169,7 +169,7 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			dat += "<a href='?_src_=prefs;preference=name;task=random'>Random Name</A> "
 			dat += "<a href='?_src_=prefs;preference=name'>Always Random Name: [active_character.be_random_name ? "Yes" : "No"]</a><BR>"
 
-			dat += "<b>Name:</b> "
+			dat += "<b>[TOOLTIP_WRAPPER("Name:", 400, "Each speices must follow a naming policy. Please read the \"Naming Guidelines\" page on the wiki and Rule 1.7.")]</b> "
```
Comment (itsmeow) [inline code/modules/client/preferences.dm]: species not speices

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
> basic html `<a href="link">Text</a>`

it doesn't work because the char pref page is a browser already, making it to go to wiki page.

**Comment by itsmeow** [comment]
Add target=_blank

**Comment by EvilDragonfiend** [comment]
> Add target="_blank"

yeah I totally forgot that

**Review by itsmeow — DISMISSED** [review]
I like it. Headmins are good with it. Would prefer if the links looked like classic underlined blue text but since it inherits the button style it's fine

**Comment by itsmeow** [comment]
You can still click on the name button right? From these screenshots it looks like it covers the button

**Comment by EvilDragonfiend** [comment]
> You can still click on the name button right? From these screenshots it looks like it covers the button

yeah it's literally tooltip and these are fine. I am thinking another format like `Name: [name] ..?` and ..? floats tooltip.

**Comment by EvilDragonfiend** [comment]
phew, I fixed it before it is merged with a critical error lol

**Comment by PowerfulBacon** [comment]
Should this be adjustable or disable able via a config for our downstreams?

**Comment by itsmeow** [comment]
I feel like if everything is a config for downstreams we are making a ton of work for ourself when it can just be a tiny code change. The wiki links as they are don't really convert well to a config value

Suppose there could be one to turn the tooltip on or off

**Comment by EvilDragonfiend** [comment]
I will make a change then let's see how it'd be.

**Comment by EvilDragonfiend** [comment]
I feel this is a bit janky, but well it's pretty working well.

**Comment by itsmeow** [comment]
Actually very impressed

File: code/__HELPERS/html_assistant.dm [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/proc/get_wiki_url()
+	return CONFIG_GET(string/wikiurl)
```
Comment (itsmeow) [inline code/__HELPERS/html_assistant.dm]: this seems a weird place to put this proc, like it would belong more next to the config definition itself maybe

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/__HELPERS/html_assistant.dm (hunk 4 above)
Comment (EvilDragonfiend) [reply code/__HELPERS/html_assistant.dm]: hmm, I don't see any better helpers among what we have for this

File: code/modules/client/preferences.dm (hunk 1 above)
Comment (Crossedfall) [reply code/modules/client/preferences.dm]: Probably a bit too specific without this being a config entry

File: config/tooltips.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+# "you need to <a href='wiki://guide_to_chemisty'>read this guide</a>."
+# if your config has no wiki address, messages will not have hyperlinks.
+
+## Key and values
+# first string is a key, and later string after first whitespace is value.
+# if you have "rule2_4 please read rule2_4"
+# rule2_4 will be a key for `TOOLTIP_CONFIG_CALLER(hover_text, width, "put key here")`
+
+#### Configurable tooltips
+#naming_policy Each species must follow a naming policy. Please read the [[Naming_Guidelines "Naming Guidelines"]] page on the wiki and [[Rules#1_You_Must_Roleplay Rule 1.7]].
```
Comment (Crossedfall) [inline config/tooltips.txt]: Good implementation 👍

File: code/__HELPERS/global_lists.dm [hunk 6]
```diff
@@ -73,6 +73,37 @@
...
+			stacker += copytext(text_value, 1, opencut)       // >> "you need to
+			text_value = splicetext(text_value, 1, opencut+1) // >> [[guide_to_chemisty read this guide]] please."
+			var/spacecut = findtext(text_value, " ")
+			var/closecut = findtext(text_value, "\]\]")
+			stacker += OPEN_WIKI(copytext(text_value, 2, spacecut), copytext(text_value, spacecut+1, closecut))  // replace [[ ]] wapper in text_value to hyperlink
+			stacker += copytext(text_value, closecut+2)       // >> please."
+
+			text_value = jointext(stacker, "")    // result >> "you need to <a>read this guys</a> please."
+			opencut = findtext(text_value, "\[\[")
+		GLOB.tooltips += list("[key]" = text_value)
```
Comment (PowerfulBacon) [inline code/__HELPERS/global_lists.dm]:
```suggestion
		GLOB.tooltips[key] = text_value
```

**Comment by EvilDragonfiend** [comment]
changed key to track where it belongs to easily.

File: code/__HELPERS/html_assistant.dm (hunk 4 above)
Comment (itsmeow) [reply code/__HELPERS/html_assistant.dm]: why is this even a proc

Comment (EvilDragonfiend) [reply code/__HELPERS/html_assistant.dm]: it had indentation errors when I used GONFIG_GET before. I tried it again and now it is fine. hmm...

**Comment by EvilDragonfiend** [comment]
I enhanced the feature since it can be used from somewhere too. `encode_wiki_link` is a thing now.
also, it tracks when a case when there's no custom title. for example `[[chemistry_guide]]` will be `chemistry guide` with the hyperlink

**Review by itsmeow — DISMISSED** [review]
Amazing

File: code/__DEFINES/html_assistant.dm [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+</style>"
+// IE11 does not support the max-content attribute, so 'width: max-content;' doesn't work.
+
+#define TOOLTIP_WRAPPER(hover_me, width_px, tooltip_text) \
+"<div class='tooltip'>[hover_me]<span class='tooltiptext' style='width: [width_px]px'>[tooltip_text]</span></div>"
+
+#define TOOLTIP_CONFIG_CALLER(hover_me, width_px, config_key) \
+"[(GLOB.tooltips[config_key] ? "<div class='tooltip'>[hover_me]<span class='tooltiptext' style='width: [width_px]px'>[GLOB.tooltips[config_key]]</span></div>" : "[hover_me]")]"
+
+#define OPEN_WIKI(wiki_url, text) (CONFIG_GET(string/wikiurl) ? "<a href='[CONFIG_GET(string/wikiurl)+"/"+wiki_url]' target='_blank'>[text]</a>" : "[text]")
```
Comment (PowerfulBacon) [inline code/__DEFINES/html_assistant.dm:34]: Does this open a new window or attempt to embed it inside the local UI which will probably break?

Comment (EvilDragonfiend) [reply code/__DEFINES/html_assistant.dm:34]: `target='_blank'`

Comment (EvilDragonfiend) [reply code/__DEFINES/html_assistant.dm:34]: I mean it opens a new IE browser properly. checked in local.

Comment (itsmeow) [reply code/__DEFINES/html_assistant.dm:34]: "New IE"

as in the system browser... or a game IE browser? Because it should be the system browser

Comment (EvilDragonfiend) [reply code/__DEFINES/html_assistant.dm:34]: a game IE browser. I see nothing bad from it though?

Comment (EvilDragonfiend) [reply code/__DEFINES/html_assistant.dm:34]: I mispoke. It is the IE11 in windows, not byond connected browser.

**Comment by itsmeow** [comment]
cries i was about to merge this

**Comment by EvilDragonfiend** [comment]
solved

**Review by itsmeow — APPROVED** [review]

**Review by Crossedfall — APPROVED** [review]
Config seems fine.

**Pull request merged**